### PR TITLE
Fixes #147 - Disable existing admin backend and trust LTI for users that are admin

### LINKED ISSUE #147: Disable old admin backend and trust LTI for users that are admin
We currently have a backend login for admin that we're using in the LTI version. When in LTI this should be disabled and there should be some link displayed for users who are super users in the database to be able to click on and access the admin interface. 

This link may need to be added to plasmic or maybe we can just put it in the corner. We currently don't have any menus in this app but maybe it could use a ☰ left of the "Canvas App Explorer" title and admin could be in there. Not sure what else though?

### PR DESCRIPTION
**Pull request by jonespm**
Because the backdoor to add an initial pylti and super user are removed from this. I'm planning on also updating the Wiki documentation. This significantly simplifies local setup.

### Run this command to create an initial PyLTI Key. These are the minimal arguments

These are the minimal values to set. You can run it multiple times to update a previous client_id+issuer pair. To remove you'll have to go into the admin UI or database.

`docker exec -it canvas_app_explorer ./manage.py manage_pylti --tool_key="LTI Key" --title="Canvas Test" --client_id=85530000000000126`

You can additionally specify 
`--platform` (Defaults to canvas.instructure.com)
`--deployment_ids` (Needs to be set unless the .env value LTI_CONFIG_DISABLE_DEPLOYMENT_ID_VALIDATION=True is set)
  You can specify multiple like 
  `--deployment_ids "303:43090cc0c11063fdf5b270094da05ed6e23ea6ad" "100:43090cc0c11063fdf5b270094da05ed6e23ea6ad"`
  These are obtained from the External Apps Page on Canvas

### After you set this up, login from Canvas and Authorize the app. You should see the home page.

Then run this to promote yourself or another logged in user initially to staff. (Fill in the UPDATE_USERNAME= part)

`UPDATE_USERNAME="jonespm@umich.edu";docker exec -it canvas_app_explorer ./manage.py shell -c "from django.contrib.auth.models import User; User.objects.update_or_create(username='$UPDATE_USERNAME', defaults=dict(is_superuser=True, is_staff=True))"`

After the first time doing this you can login as a user and promote through the admin UI and the link.

### CODE REVIEW

**Comment by jonespm** [comment]
#212 should be merged or rejected before this is merged as this includes that commit.

File: backend/urls.py
Comment (ssciolla) [inline backend/urls.py:39]: Is there a way to remove these rather than overriding them? Just curious. I saw you could customize the `AdminSite` class, but didn't see how Django itself recommends doing this. https://docs.djangoproject.com/en/4.0/ref/contrib/admin/#adminsite-objects

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (zqian) [inline backend/canvas_app_explorer/management/commands/manage_pylti.py:17]: should those canvas urls be configurable for test vs production Canvas server urls?

File: backend/views.py
Comment (zqian) [reply backend/views.py]: maybe instead of name get_login_template, can we use get_error_template instead, and pass in error message as parameter?

File: backend/urls.py
Comment (jonespm) [reply backend/urls.py:39]: I didn't see any ideas either. this [admin.site.urls](https://github.com/django/django/blob/c627226d05dd52aef59447dcfb29cec2c2b11b8a/django/contrib/admin/sites.py#L317) is a dynamic method that generates urls from get_urls. 

It would be possible to override/extend some of these classesthat but I don't know if that's a better idea than this.

The URLS [are here.](https://github.com/django/django/blob/c627226d05dd52aef59447dcfb29cec2c2b11b8a/django/contrib/admin/sites.py#L269)

There are some others here like /admin/password-change but they redirect to /admin/login first.

I basically [used this suggestion from here](https://stackoverflow.com/a/61946707/3708872) without finding a better one. 

I think the eventual fix would be that we write our custom admin, I just wanted to remove any potential backdoors.

Comment (jonespm) [reply backend/urls.py:39]: There were some other ideas here. Rather than define a template with a nice message we **could** use the view [django.views.defaults.page_not_found](https://docs.djangoproject.com/en/4.0/ref/views/#the-404-page-not-found-view).

I also didn't try if it's possible to do something like add the path
`    path('admin/', admin.site.urls),`

But then remove these paths from the list?

File: README.md
Comment (zqian) [inline README.md:14]: Since the admin role is solely defined in CAE Django database, we may need to add an instruction line here. 

Can include a sql update statement too: `update auth_user set is_staff=1 where id=x;`

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (ssciolla) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:17]: Not sure if this is exactly what @zqian is saying, but I was thinking you could just have the user provide the platform URL, and have the endings be configurable via environment variables with good defaults. I think this is fine, but it would make it a lot easier for test configuration if you just had to provide `--platform=canvas.test.instructure.com` and instead of entering all these `_url` flags on the command line.

File: README.md
Comment (jonespm) [reply README.md:14]: I put this in the instructions in the notes how you can update this with the application. That's **possibly** better than requesting the user login and run SQL. Basically is the same action. I figured this would just be in the wiki. 

`UPDATE_USERNAME="jonespm@umich.edu";docker exec -it canvas_app_explorer ./manage.py shell -c "from django.contrib.auth.models import User; User.objects.update_or_create(username='$UPDATE_USERNAME', defaults=dict(is_superuser=True, is_staff=True))"`

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:17]: All of these are optionally configurable.

It doesn't **seem** to matter on test/dev/prod which ones you use. You probably should use the one for your server's URL but it works for me even with these defaults, at least on dev. 

If it doesn't work you can add these command line arguments and run it again and it will update it. 

Yeah, maybe rather than making these all individually configurable (or optionally in addition to) we could just have like a platform URL as you suggest. 

Actually that's probably easier for this really and a good idea. I'll just remove all those and if `--platform` is defined we can use that, otherwise just use `canvas.instructure.com`.

Comment (ssciolla) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:17]: I think dev uses `canvas.instructure.com` for the platform; not sure if that works for test or beta though.

Comment (zqian) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:17]: This canvas.instructure.com setting also works for Canvas test server. I tested it with the integration of my localhost.

I am confused now. Is the platform value already remains unchanged as canvas.instructure.com?

Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:17]: I haven't done extensive testing but I think the only thing that doesn't work and return the same results is the jwks URL like this https://canvas.instructure.com/api/lti/security/jwks

But we aren't using this anyway in this app so it doesn't matter. **Possibly** also the OAuth URL's but those are in the .env file. 

I think there's a chance that canvas.instructure.com could be down while our instance is still up and break our app. So for prod and test we probably want to use the domain matching the instance we're testing on.

Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:17]: I'm simplifying all of this so you only have to pass platform. 

You can also go into the database and edit these too, but I've found it to be a lot easier to run a management command to get this provisioned that editing them in the db, especially for the first one.

Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py]: I updated the command to just take the platform arg

File: backend/views.py
Comment (jonespm) [reply backend/views.py]: Sure, I can change the name and pass it from here if I can't get a generic template to work better.

File: backend/urls.py
Comment (ssciolla) [reply backend/urls.py]: I think what you're doing is probably OK. Changing the name of the view was probably a good call.

Comment (jonespm) [reply backend/urls.py]: I looked into this and couldn't figure out anything great.

I have an idea of running it through a configurable filter. This **feels** like is should work. This produces a filtered tuple that looks like we can pass it into path. This can for sure be improved. 

```
from django.urls import URLPattern

def exclude_pattern(var, patterns: list):
    # Just ignore these if they don't have a name because it's probably a URLResolver
    if type(var) is URLPattern and var.name in patterns:
        return False
    return True

(list(filter(partial(exclude_pattern, patterns=["login", "logout", "password_change", "password_change_done"]), admin.site.urls[0])), admin.site.urls[1], admin.site.urls[2])
```

This looks like it's always a pattern of  [(list, tuple)](https://github.com/django/django/blob/c627226d05dd52aef59447dcfb29cec2c2b11b8a/django/urls/conf.py#L71).


I didn't like the default look of the page_not_found and it's not **really** a 404.

**Review by zqian — APPROVED** [review]

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (ssciolla) [inline backend/canvas_app_explorer/management/commands/manage_pylti.py:27]: Maybe "LTI Tool Key record"?

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (ssciolla) [inline backend/canvas_app_explorer/management/commands/manage_pylti.py:65]: `options['deployment_ids']` needs to be wrapped in a `json.dumps(...)` here. It didn't work for me when I used a single deployment ID, with or without wrapping `"`s around it (they get stripped by the parser I think).

File: backend/urls.py
Comment (ssciolla) [inline backend/urls.py:36]:
```suggestion
    # Redirect all of these to an error template page to display a message
```

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (ssciolla) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:27]: Some other messages about the other steps may be appropriate.

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (ssciolla) [inline backend/canvas_app_explorer/management/commands/manage_pylti.py:51]: I'm not sure how computationally expensive `RSA.generate` or these other methods are (it does seem to hang for a second), but it might make sense to use another pattern so you don't need to call these methods if it already exists. Something like this:
```
try:
    lti_key = LtiToolKey.objects.get(name=options['tool_key])
except LtiToolKey.DoesNotExist:
    key = RSA.generate(4096)
    lti_key = LtiToolKey.objects.create(
      name=options['tool_key'],
      private_key=key.exportKey().decode('utf-8'),
      public_key=key.publickey().exportKey().decode('utf-8')
    )
```
There's also a question of whether to use `queryset.exists()` or `queryset.get` with a `try`/`except`. Either probably is fine here, IDK.

File: backend/urls.py
Comment (ssciolla) [reply backend/urls.py:42]: Nit-pick: You could also put periods at the end of the sentences.

**Review by ssciolla — CHANGES_REQUESTED** [review]
Sorry for all the comments. I think I'd have given approval if the `deployment_ids` flag was functional, but there may be some other suggestions here you'll find worth considering. Let me know if you need clarification or disagree with anything.

File: requirements.txt
Comment (ssciolla) [reply requirements.txt:30]: Okay, that's fine, we can stick with this one.

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (ssciolla) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:65]: FYI, this is the only comment that blocked me from approving. I'll approve if you resolve it.

File: backend/urls.py
Comment (jonespm) [reply backend/urls.py:42]: All of these views could _probably_ have the same message. I put one in as the default. 

Or do you mean just to pass a partial like "Account Login" and "Admin Login" rather than the entire message? The get_error_template *might* be able to get the route then it'd doesn't need to be passed in separately and display like

`accounts/login is disabled as you need to launch this tool from an LTI tool provider.` I tried using default view like 
`defaults.page_not_found` but that would require us to provide a better 404 template page, otherwise there's no information on it. That seemed a little out of scope.

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:66]: I was thinking of having some of this semi self contained so I could contribute it back to the pylti project. But this is really Canvas specific and we'd need different patterns for other LMSes. This was really just intended as an easier way than running SQL queries or editing the DB with an IDE to get this initial row in here that didn't currently exist. I was trying to figure this out for MyLA too so we could move the configuration there to the DB, but MyLA not as likely to be multi-tenant as this app will hopefully someday be. 

The only semi-annoying part of doing it without the script is generating the RSA keys. MyLA does have a utility for that but I usually just use the IMS website. I was just hoping this would save some time for local dev.

File: backend/views.py
Comment (jonespm) [reply backend/views.py:14]: I think this is simple enough to just leave it as a function based view to be consistent with the other function.

We may want to have a separate issue where we convert these to class based views if these are getting complicated. 

Using the `messages` is probably unnecessary here too though. Really I'm starting to lean toward the fix I came up with earlier to remove these paths entirely from the start and remove this error code entirely.

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:5]: I could see renaming it to create_pylti_tool_config. It really is specific to this library. 

I originally envisioned it as a command that could be used to CRUD the records in these PyLTI tables but right now it just can Create and Update.

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:27]: I can just drop the table part, will add a few other messages

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:51]: I'm rewriting it with this pattern.

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py]: Okay, this looks like it was saving it with single quotes instead of double. I switched to just use json.dumps as you suggested.

File: backend/urls.py
Comment (jonespm) [reply backend/urls.py]: So this seemed like a good idea but removing these results in an error when other code in admin tries to reverse these

`Reverse for 'login' not found. 'login' is not a valid view function or pattern name.`

[This solution to just wrap the method login in a login_required](https://stackoverflow.com/a/13186337/3708872) seems to work fine.

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py]: I think this could be done later if we write a test for this, but I'd like to try to get it back upstream. Just going to resolve for now. I refactored this since this change too.

File: backend/urls.py
Comment (jonespm) [reply backend/urls.py]: Removed this in 7cc0bc93a4a708751d317e604719f4e92dc3785a

File: backend/views.py
Comment (ssciolla) [inline backend/views.py:2]:
```suggestion
```

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (ssciolla) [inline backend/canvas_app_explorer/management/commands/manage_pylti.py:5]: I think `RsaKey` and `IntegrityError` are unused?

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (ssciolla) [inline backend/canvas_app_explorer/management/commands/manage_pylti.py:12]: Please check this help string for accuracy, clarity. Sounds initially like it just generates the key pair, but it does more than that.

File: backend/urls.py
Comment (ssciolla) [inline backend/urls.py:35]: So a couple things I noted, some of which may be commented on in the SO thread:
1) This doesn't disable `logout`, but it doesn't break if you use it on the Admin site. I don't see this as a huge issue, but thought I'd raise it as a discussion point.
2) You can still somewhat indirectly surface the login page via redirect. You can observe this if you login as a non-admin user, view the frame source, then navigate to `/admin`. You are logged in, so it doesn't block you from visiting the page, but it knows you don't have the proper permissions and lets you log in as a different user.

**Review by ssciolla — APPROVED** [review]
I'm approving, as I don't think we need another full review cycle, but I'm providing some additional comments for consideration.

File: backend/canvas_app_explorer/management/commands/manage_pylti.py
Comment (jonespm) [reply backend/canvas_app_explorer/management/commands/manage_pylti.py:10]: Thanks, I re-read this and updated it to describe a little more what it's doing.

File: backend/urls.py
Comment (jonespm) [reply backend/urls.py:35]: Yeah, I noticed that if you call that it redirects to login which is disabled. 

If someone did figure out the admin and tried to login, they wouldn't be able to use the page. I don't think this is a risk or anything that's going to be mentioned. It's mostly to just block any potential external login attempts. 

It looks like I probably could just use user_passes_test instead to block it entirely.

`admin.site.login = user_passes_test(False, admin.site.login)`

**Pull request merged**
